dhall/src/Dhall/ETags.hs

```diff
@@ -0,0 +1,161 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | This module contains the implementation of the @dhall tags@ command
+
+module Dhall.ETags
+    ( -- * ETags
+      generate
+    ) where
+
+import Control.Monad
+import Control.Exception
+import Data.List (isSuffixOf)
+import Data.Semigroup (Semigroup(..))
+import Data.Text (Text)
+import Dhall.Util (Input(..))
+import Dhall.Core (Expr(..), Binding(..))
+import Dhall.Src (Src(..))
+import Dhall.Parser
+import System.Directory ( doesDirectoryExist
+#if MIN_VERSION_directory(1,3,0)
+                        , pathIsSymbolicLink
+#elif MIN_VERSION_directory(1,2,6)
+                        , isSymbolicLink
+#endif
+                        , getDirectoryContents
+                        )
+import System.FilePath ((</>), takeFileName)
+import Text.Megaparsec (sourceLine, sourceColumn, unPos)
+
+import qualified Data.Map      as M
+import qualified Data.Text     as T
+import qualified Data.Text.IO  as TIO
+import qualified Dhall.Map     as DM (keys)
+
+data FilePos = FP
+    { fpLine :: Int
+    , fpOffset :: Int
+    } deriving (Eq, Ord)
+
+newtype Tags = Tags (M.Map FilePath (M.Map FilePos Tag))
+
+instance Semigroup Tags where
+    (Tags ts1) <> (Tags ts2) = Tags (M.unionWith M.union ts1 ts2)
+
+data Tag = Tag
+    { tagDefinition :: Text
+    , tagName :: Text
+    }
+
+{-| Generate ETags for Dhall expressions
+
+    Arguments:
+
+    * Path to directory or filename that should be indexed
+    * List of suffixes for dhall files [".dhall"] by default
+    * Flag if generate should follow symlinks
+-}
+generate :: Input -> [Text] -> Bool -> IO Text
```

Comment:
Is the resulting `Text` valid unicode? If it isn't, we should probably use `ByteString`. In that case we should use bytestring `Builder`s to concatenate the final representation.

Reply:
Dhall reads it as unicode. As I see, all modules are working with dhall files as they are valid unicode.
So I think, it's safe enough to think, it's valid unicode.
In case, I miss something, can you show me some example how dhall works with non-unicode files, please?

Reply:
Dhall files are indeed UTF8

I was thinking about the tags-files we're producing, like dhall/tests/etags/assertB.tags, which looked a bit funny. But I checked it with `isutf8` and it seems to be fine. Never mind then. :)